[RFC] Use pre-compiled headers to speed up LLVM build (~1.5-2x)

aengelke opened the pull request:
Building LLVM is slow and dominated by the C++ front-end. Most time is spent in (repeatedly) parsing headers. C++ modules build didn't really help, is fragile, and kills parallelism. Therefore, I propose to use PCH for frequently used headers (i.e., C++ stdlib, Support, IR, CodeGen). There shouldn't be much difference for incremental compilation, as many headers I put into the PCHs transitively get included into most source files anyway.

Time breakdown in seconds for building libLLVM.so (collected with -ftime-trace, -O1, -DLLVM_TARGETS_TO_BUILD="X86;AArch64"):

```
Phase                           main main+TPDE+PCH
------------------------------- ---- -------------
ExecuteCompiler                 9913          4084
Frontend                        7842          2773 (PCH difference)
  Source                        5471          1213
  PerformPendingInstantiations  1837           916
  CodeGen Function               248           278
Backend                         2014          1256
  Optimizer                     1332          1243 (probably just measurement noise)
  CodeGenPasses                  675             3 (TPDE fallbacks to LLVM for 10 CUs)
  TPDE                             0             7

wall-time [s] (48c/96t)       126.48         66.04
(AMD EPYC 9454P, 2.75-3.80 GHz, 372 GiB memory)
```
(Edit: new data with idle machine + assertions disabled, individual results are still a bit noisy due to hyper-threading; no substantial change in relative factors.) (NB: on this machine, much larger wall-time reductions are rather unlikely: SLPVectorizer.cpp takes >40s alone.)

The back-end is not really relevant here, but enabling PCH (this PR) reduces the front-end time by 2x. On [c-t-t, this shows up as a 30% wall-time/instructions reduction in stage2-clang](https://llvm-compile-time-tracker.com/compare.php?from=937009a76c6af9799ccbf89d7d8a07dbb4917f86&to=8e38a6f850d685a437a5bf6a8823c7996ba484bd&stat=instructions:u). Further improvements for Clang should also be possible (clang/AST/Decl.h is expensive with 1219s), but is not trivial due to the use of object libraries.

I'm opening this PR for early feedback/direction before spending more time on this:

- Does this extensive PCH use has a reasonable chance of getting merged?
  - How to deal with newly occuring name collisions? (e.g. llvm::Reloc from llvm/Support/CodeGen.h vs. lld::macho::Reloc from lld/MachO/Relocations.h; where e.g. lld/MachO/InputSections.cpp uses both namespaces)
- Enable by default vs. not? E.g., CI should probably not use this to catch missing includes, but CI should also keep it working as PCHs can lead to new errors.
- PCH reuse is currently hard-coded in llvm_add_library, how to make this more elegant?

cc @nikic @rnk @boomanaiden154 (not sure who else is interested in LLVM build times + looking at CMake)

aengelke commented:
FWIW, statistics about optimization passes >5s (no adaptors, wrappers, etc.) on libLLVM (NB: pass times can include analyses):
```
   Total   NumRuns       PerRun  Pass
   5.01s    160476      31.19us  LoopUnrollPass
   5.35s   4404446       1.21us  TargetIRAnalysis
   6.94s   4403527       1.58us  AssumptionAnalysis
   7.10s    433061      16.41us  LoopRotatePass
   7.91s      3909    2023.59us  CallGraphAnalysis
   9.66s      1980    4877.52us  CalledValuePropagationPass
  10.19s      1980    5145.05us  AlwaysInlinerPass
  12.22s   5587932       2.19us  PostDominatorTreeAnalysis
  12.26s   2070891       5.92us  ReassociatePass
  13.61s   2070891       6.57us  ADCEPass
  13.87s    737972      18.80us  LICMPass
  13.90s   2070891       6.71us  BDCEPass
  15.25s    432540      35.25us  LoopDeletionPass
  15.41s   4462734       3.45us  LoopSimplifyPass
  17.77s   5938233       2.99us  LoopAnalysis
  18.65s      3960    4710.34us  GlobalOptPass
  19.69s  11044561       1.78us  DominatorTreeAnalysis
  20.63s    302139      68.27us  IndVarSimplifyPass
  23.03s   2070891      11.12us  MemCpyOptPass
  25.45s    302139      84.23us  LoopIdiomRecognizePass
  25.85s   4139632       6.24us  PostOrderFunctionAttrsPass
  28.08s   2070891      13.56us  SCCPPass
  30.01s      1980   15158.71us  IPSCCPPass
  34.19s   3495375       9.78us  BranchProbabilityAnalysis
  42.01s   2843306      14.77us  MemorySSAAnalysis
  49.53s   3495375      14.17us  BlockFrequencyAnalysis
  64.37s   6536108       9.85us  SROAPass
  82.56s   4304741      19.18us  EarlyCSEPass
  91.52s  13070009       7.00us  SimplifyCFGPass
 248.60s   2069816     120.11us  InlinerPass
 279.41s  10996635      25.41us  InstCombinePass
1274.61s      1980  643743.37us  Optimizer

```

aengelke committed:
pch for unittests

aengelke commented:
Another data point: I can build LLVM (incl. tools) in 6-7 minutes on my laptop now (M2 Macbook Air, 8 cores); building unit-tests take another 2.5mins. Interestingly, with PCH optimizations are comparably expensive here:

```
ExecuteCompiler                 3272
Frontend                        1533
  Source                         476
  PerformPendingInstantiations   570
  CodeGen Function               300
Backend                         1699
  Optimizer                     1056
    InlinerPass                  264 (90.27us per run; 2934128 runs)
    InstCombinePass              163 (10.69us per run; 15292003 runs)
    SimplifyCFGPass               67 (3.77us per run; 18032531 runs)
    EarlyCSEPass                  58 (9.87us per run; 5880387 runs)
    SROAPass                      51 (5.69us per run; 9020246 runs)
    BlockFrequencyAnalysis        47 (9.41us per run; 5048403 runs)
    BranchProbabilityAnalysis     33 (6.72us per run; 5048403 runs)
    AAManager                     30 (2.57us per run; 11970407 runs)
    MemorySSAAnalysis             29 (7.63us per run; 3921385 runs)
    IPSCCPPass                    28 (12019.39us per run; 2410 runs)
    PostOrderFunctionAttrsPass    23 (3.98us per run; 5868256 runs)
  CodeGenPasses                  639

wall-time (8c/8t) ~400s
```

Most expensive CUs >15s:

```
15.769s unittests/SandboxIR/SandboxIRTest.cpp
16.174s lib/CodeGen/SelectionDAG/DAGCombiner.cpp
16.238s lib/Transforms/IPO/MemProfContextDisambiguation.cpp
16.457s unittests/Frontend/OpenMPIRBuilderTest.cpp
16.702s unittests/ADT/DenseMapTest.cpp
16.779s lib/Transforms/IPO/AttributorAttributes.cpp
16.920s lib/Target/AArch64/AArch64ISelLowering.cpp
28.694s tools/llvm-readobj/dir/ELFDumper.cpp
28.865s lib/Target/X86/X86ISelLowering.cpp
30.295s lib/Passes/PassBuilder.cpp
33.410s lib/Transforms/Vectorize/SLPVectorizer.cpp
54.741s unittests/Frontend/OpenMPDecompositionTest.cpp
```

boomanaiden154 reviewed (commented):
Not exactly sure how you need to invoke the clang driver to use precompiled headers, but it seems like sccache might not have support currently (https://github.com/mozilla/sccache/issues/615). Might still be useful for local builds, but we would have to turn it off in premerge CI until it works with caching.

Either way, I would probably like to see a proper RFC on discourse for this.

aengelke commented:
[RFC on Discourse](https://discourse.llvm.org/t/rfc-use-pre-compiled-headers-to-speed-up-llvm-build-by-1-5-2x/89345)

aengelke committed:
[CMake][LLVM] Add PCH infrastructure and LLVMSupport PCH

aengelke committed:
[CMake][IR] Add PCH

aengelke committed:
[CMake][CodeGen] Add PCH

aengelke committed:
[CMake][AST] Add PCH

aengelke committed:
[CMake][CodeGen] Add PCH for Clang CodeGen

aengelke commented:
Thanks for checking GCC, I rarely use it for building LLVM. Reducing the number of headers seems to be important for GCC, there's certainly some room for improvements. Some observations from quick testing and profiling of GCC:

- PCH are much larger than with Clang. Even after reducing the size of Support/pch.h, the Support PCH is still 132M (Clang: 24M).
- As a consequence, max-rss also increases strongly (e.g., from 397M to 751M).
- cc1plus system time increases strongly; on a random file, perf indicates >20% time spent in handling page faults (2x more than before), probably due to the large PCH.
- Parsing time goes down substantially, but, according to perf on cc1plus, most of the improvement is negated by instantiate_pending_templates -> instantiate_decl, which consumes much more time in the PCH build (on a random file, this increases from 5% to 25% with almost identical total compile time).
- -ftime-report is not useful for profiling, it strongly affects execution times. I don't know how to get more accurate information on what exactly causes GCC to slow down that much.

The warnings seem to only come from executables, most files in libraries are not afffected.

aengelke committed:
[LLD][MachO][NFC] Rename Reloc to Relocation

Due to heavy use of using namespace llvm, Reloc is often ambiguous with
llvm::Reloc, the relocation model. Previously, this was sometimes
disambiguated with macho::Reloc. This ambiguity is even more problematic
when using pre-compiled headers, where it's no longer "obvious" whether
it should be Reloc or macho::Reloc.

Therefore, rename Reloc to Relocation. This is also consistent with
lld/ELF, where the type is also named Relocation.

pinskia commented:
Llvm CPU time: 5497.8 usr, 297.6 sys
Vs
Gcc 5232.6 usr, 559.8 sys

Hmm. So from the looks of it clang/llvm could be optimized more. Gcc is most likely spending more time in system as it writes out the .s file and then gas reads it back in and assembles it.
So gcc is spending less time in user space to do work, ~5% if my math is correct. But 2x as much time in the kernel. Which does corresponding to writing out a .s file.
